feat: expose public interface for locating Chrome installations

knksmith57 opened the pull request:
Add `Launcher.getInstallations()` public static method that returns an
array of paths to available Chrome binaries. Also, refactor
`Launcher#launch()` to use it.

`Launcher.getInstallations()` uses `chrome-finder` under the hood, so all
previous resolution logic is preserved:
- if `CHROME_PATH` env variable is set, it will be included
- if a Chrome Canary is detected, it will be included
- if a Chrome (stable) is detected, it will be included

Resolves #176

knksmith57 commented:
@googlebot I signed it!

knksmith57 commented on src/chrome-launcher.ts:
```diff
@@ -167,6 +167,10 @@ class Launcher {
   }
 
+  static getInstallations() {
+    return chromeFinder[getPlatform() as SupportedPlatforms]();
```
note: `await` was previously used here; however, afaict, `chromeFinder[]()` methods intentionally use blocking I/O and synchronous interfaces --I opted to implement `getInstallations()` as _non_-async to avoid any ambiguity to the caller as to whether or not the method was blocking.

happy to change this if y'all feel differently 😄

patrickhulce commented on src/chrome-launcher.ts:
```diff
@@ -205,7 +209,7 @@ class Launcher {
     }
     if (this.chromePath === undefined) {
-      const installations = await chromeFinder[getPlatform() as SupportedPlatforms]();
+      const installations = (<typeof Launcher>this.constructor).getInstallations();
```
can't we just call `Launcher.getInstallations()`? :)

knksmith57 replied:
wow you were fast! thanks for hopping in so quickly!

I was proactively typing up a note for why I opted for this but was obvs too slow lol --I'll respond back in the other thread.

knksmith57 replied:
addressed in other thread --closing out :)

knksmith57 commented on src/chrome-launcher.ts:
```diff
@@ -205,7 +209,7 @@ class Launcher {
     }
     if (this.chromePath === undefined) {
-      const installations = await chromeFinder[getPlatform() as SupportedPlatforms]();
+      const installations = (<typeof Launcher>this.constructor).getInstallations();
```
in case it's not obvious, because `Launcher` is publicly exported, rather than reference `Launcher.getInstallations()`, I opted to use late static binding here so that consumers that wish to override `getInstallations()` can.

patrickhulce replied:
I'm not sure what value that really has. We already provide `chromePath` as an option and we'd never do anything with installations past the first one.

knksmith57 replied:
that's a super fair and valid point.

I suppose my inclination is to lean towards ensuring that public class methods can be overridden by consumers of the library --but you've made a great point that, currently, the only call-site that uses this already supports being overridden via another interface.

I'll replace with `Launcher.getInstallations()` 😄

knksmith57 replied:
done!

patrickhulce commented on test/chrome-launcher-test.ts:
```diff
@@ -133,6 +133,11 @@ describe('Launcher', () => {
   });
 
+  it('searches for available installations', async () => {
+    const installations = Launcher.getInstallations();
```
assert that we found one too? :)

knksmith57 replied:
I wasn't sure if that was safe given we don't currently require chrome as a dev dependency (that I can tell), so CI might not pass?

knksmith57 replied:
OK I added the check --hopefully that works as intended 😆

knksmith57 commented on test/chrome-launcher-test.ts:
```diff
@@ -133,6 +133,11 @@ describe('Launcher', () => {
 
+  it('searches for available installations', async () => {
+    const installations = Launcher.getInstallations();
+    assert.ok(Array.isArray(installations));
```
this test case is super weak sauce. very much open to, and would appreciate, suggestions for how to better test this feature.

as it stands, this coverage at least ensures that the method exists, is callable, and returns an array --which is better than nothing, but fails to ensure that the appropriate resolution logic is actually used.

I struggled with how to go about testing this method for a few reasons:
1. the project doesn't appear to include any module mocking dependencies and I didn't feel comfortable pulling one in without first reaching consensus that I should
2. all of the heavy lifting is done by chrome-finder, while platform resolution logic is handled by utils; short of asserting that `getPlatform()` and `chromeFinder[]()` are invoked (which likely requires a mocking lib ^^), I'm not sure what else to assert.

knksmith57 commented:
OK looks like adding that additional assert worked (I didn't realize we codified the chrome dep in our travis config; I should have looked at that first 😅).

I think I've addressed all feedback --thanks again for the _super_ fast review!

patrickhulce commented:
yup the rest of our tests would fail if we didn't have chrome available so it's a safe bet :)

patrickhulce commented:
looks great, thanks again @knksmith57! 🎉 💯

paulirish commented:
great idea :D
